### ISSUE TITLE
Consider structures other than `Array[Byte]`

### ISSUE THREAD

**Issue by marklister** [marklister/base64#8]
From #6 (@bblfish)

Perhaps there is an extra improvement you can add. Is it possible to abstract somewhat from an array? For example the JS seems to return a [ArrayBuffer](https://github.com/scala-js/scala-js/blob/master/library/src/main/scala/scala/scalajs/js/typedarray/ArrayBuffer.scala) which I think can be turned into a java.nio.ByteBuffer using [TypedArrayBufferBridge](https://github.com/scala-js/scala-js/blob/master/library/src/main/scala/scala/scalajs/js/typedarray/TypedArrayBufferBridge.scala).

So it would be useful to have a method that could work with those elements and not just arrays. I am not yet quite sure what the best abstraction here would be... Certainly given that you don't need to change the source object, an Array is too precise.

There is some example code here using a ByteBuffer https://gist.github.com/mseddon/1cfcb0970272cac40497

**Comment by marklister** [comment]
How does this look?

```scala
scala> import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.ArrayBuffer

scala> val a= ArrayBuffer[Byte]() ++ "abc".getBytes
a: scala.collection.mutable.ArrayBuffer[Byte] = ArrayBuffer(97, 98, 99)

scala> a.toBase64
res0: String = YWJj
```

It works on a `Seq[Byte]`

**Comment by bblfish** [comment]
That probably helps in many situations. But I don't know if it helps with `java.nio.ByteBuffer`, other than having to convert that to that to an ArrayBuffer. Is there a way to do that efficiently? ( ie, really one wants to be allocating no new memory )

**Comment by marklister** [comment]
The latest commit (and I published it to bintray) supports a `Seq[Byte]` `ArrayBuffer [Byte]` etc.

java.nio.ByteBuffer has an `array()` method.  @bblfish, if I read your initial post correctly you're deriving ByteBuffer from ArrayBuffer so I don't think we need to support directly?

**Comment by bblfish** [comment]
yes, no need to support ArrayBuffer as that is a JS view on something that is very similar to `java.nio.ByteBuffer`. I suppose that is why there is a transformation in ScalaJS from ArrayBuffer to a `java.nio.ByteBuffer` since [scala-js 0.6.1](http://www.scala-js.org/news/2015/03/03/announcing-scalajs-0.6.1/). 

I can try the `array()` method on `ByteBuffer` though my guess is that is not the most efficient way to do things. For my purposes efficiency is not paramount at the moment. But as you start working on the efficiency of your library you'll want to use its features directly. It should not be too hard then to get something extremely efficient.  perhaps there is something to learn from @mseddon 's  https://gist.github.com/mseddon/1cfcb0970272cac40497

**Comment by bblfish** [comment]
This is my current code

```scala
   import com.github.marklister.base64.Base64._
   val bb = TypedArrayBuffer.wrap(sig.asInstanceOf[ArrayBuffer])
   val arraybuf: Array[Byte] = new Array[Byte](bb.remaining())
   bb.get(arraybuf)
   val hashedSig = arraybuf.toBase64
```

`TypedArrayBuffer.wrap(sig.asInstanceOf[ArrayBuffer])` returns a `java.nio.ByteBuffer` .
I am not sure if there are read only `ByteBuffer`s. If so you could write a wrapper that wraps those using the few lines of code above, and that could be used like this:

```scala
bb.toBase64
```

Then later when you have time you could optimise away the copy.

**Comment by marklister** [comment]
So would

    sig.asInstanceOf[ArrayBuffer].toBase64

not work?

**Comment by bblfish** [comment]
Well that would make it a pure scala-js function. `ArrayBuffer`'s full signature is `scala.scalajs.js.typedarray.ArrayBuffer`, and I don't think you need to write something that is scala-js specific like that. Libraries should be producing the `nio` classes.

**Comment by marklister** [comment]
Ah!  I see!  I had it confused with 
http://www.scala-lang.org/api/current/index.html#scala.collection.mutable.ArrayBuffer

**Comment by bblfish** [comment]
my fault. Should have included the imports. Sorry!

**Comment by bblfish** [comment]
Got my code working! I had a bug on the server with encoding/decoding BigIntegers from hexadecimal. https://github.com/read-write-web/rww-play/commit/de8655e2b0a12d4dfe2d4c4657b06df3d9fa9a13

**Comment by marklister** [comment]
Yeah, I bumped into this same problem a few years ago. See my basen lib.
